### ISSUE TITLE
Application response sent from the POD goes missing.

### ISSUE THREAD

**Issue by suryaveerc** [kubernetes/kubernetes#18933]
I have installed kubernetes on openstack. The setup has one master and one node on coreos.

I have a pod hosting an SIP application on UDP port 5060 and I have created service as **NODEPORT** on 5060.

**The spec:**
 
    "spec": {
        "ports": [
          {
            "port": 5061,
            "protocol": "UDP",
            "targetPort": 5060,
    	"nodeport": 5060,
    	"name": "sipu"
          }
        ],
        "selector": {
          "app": "opensips"
        },
    	"type": "NodePort"
      }
**IPs**

 - Public IP of node: **192.168.144.29**.
 - Private IP of node: **10.0.1.215.**.
 - IP of the container:**10.244.2.4**. 
 - docker0 interface: **10.244.2.1**.

Now, the problem. 
I send a SIP request to the application and expect a 200 OK response, which  I am not getting.

To trace the same, I installed TCPDUMP on the container and the node.
On the container, I can see the request and response packet captured while on the node itself I just see the request packet. Don't understand why the packet is getting lost.

**Below is tcpdump of container:**

    tcpdump: listening on eth0, link-type EN10MB (Ethernet), capture size 1514 bytes
    06:12:20.391171 IP (tos 0x0, ttl 64, id 13372, offset 0, flags [DF], proto UDP (17), length 1010)
        10.244.2.1.50867 > 10.244.2.4.5060: [bad udp cksum 0x1ddc -> 0xe738!] SIP, length: 982
            PUBLISH sip:service-1@opensipstest.org SIP/2.0
            Via: SIP/2.0/UDP 192.168.144.10:5060;branch=z9hG4bK-5672-1-0
            Max-Forwards: 20
            From: service <sip:service-1@opensipstest.org>;tag=1
            To: sut <sip:service-1@opensipstest.org>
            Call-ID: 1-5672@192.168.144.10
            CSeq: 1 PUBLISH
            Expires: 3600
            Event: presence
            Content-Length:   607
            User-Agent: Sipp v1.1-TLS, version 20061124
    
            <?xml version="1.0"?>
            <deleted presence xml to reduce size>
    
    06:12:20.401126 IP (tos 0x10, ttl 64, id 11888, offset 0, flags [DF], proto UDP (17), length 427)
        10.244.2.4.5060 > 10.244.2.1.5060: [bad udp cksum 0x1b95 -> 0xeddc!] SIP, length: 399
            SIP/2.0 200 OK
            Via: SIP/2.0/UDP 192.168.144.10:5060;received=10.244.2.1;branch=z9hG4bK-5672-1-0
            From: service <sip:service-1@opensipstest.org>;tag=1
            To: sut <sip:service-1@opensipstest.org>;tag=332ff20b76febdd3c55f313f3efc6bb8-ca08
            Call-ID: 1-5672@192.168.144.10
            CSeq: 1 PUBLISH
            Expires: 3600
            SIP-ETag: a.1450478491.39.2.0
            Server: OpenSIPS (1.8.4-notls (x86_64/linux))
            Content-Length: 0
    
    
    06:12:20.401364 IP (tos 0x0, ttl 64, id 13374, offset 0, flags [DF], proto UDP (17), length 427)
        10.244.2.1.58836 > 10.244.2.4.5060: [bad udp cksum 0x1b95 -> 0x1bcc!] SIP, length: 399
            SIP/2.0 200 OK
            Via: SIP/2.0/UDP 192.168.144.10:5060;received=10.244.2.1;branch=z9hG4bK-5672-1-0
            From: service <sip:service-1@opensipstest.org>;tag=1
            To: sut <sip:service-1@opensipstest.org>;tag=332ff20b76febdd3c55f313f3efc6bb8-ca08
            Call-ID: 1-5672@192.168.144.10
            CSeq: 1 PUBLISH
            Expires: 3600
            SIP-ETag: a.1450478491.39.2.0
            Server: OpenSIPS (1.8.4-notls (x86_64/linux))
            Content-Length: 0

**And tcpdump of node:**

    tcpdump: listening on eth0, link-type EN10MB (Ethernet), capture size 1514 bytes
    06:12:20.390772 IP (tos 0x0, ttl 127, id 20196, offset 0, flags [none], proto UDP (17), length 1010)
        192.168.144.10.5060 > 10.0.1.215.5060: [udp sum ok] SIP, length: 982
            PUBLISH sip:service-1@opensipstest.org SIP/2.0
            Via: SIP/2.0/UDP 192.168.144.10:5060;branch=z9hG4bK-5672-1-0
            Max-Forwards: 20
            From: service <sip:service-1@opensipstest.org>;tag=1
            To: sut <sip:service-1@opensipstest.org>
            Call-ID: 1-5672@192.168.144.10
            CSeq: 1 PUBLISH
            Expires: 3600
            Event: presence
            Content-Length:   607
            User-Agent: Sipp v1.1-TLS, version 20061124
    
            <?xml version="1.0"?>
            <presence xmlns="urn:ietf:params:xml:ns:pidf" />

Nodeport rules from Iptable

    Chain KUBE-NODEPORT-CONTAINER (1 references)
     pkts bytes target     prot opt in     out     source               destination
       12  8622 REDIRECT   udp  --  *      *       0.0.0.0/0            0.0.0.0/0            /* default/opensips:sipu */ udp dpt:5060 redir ports 40482
        3    95 REDIRECT   udp  --  *      *       0.0.0.0/0            0.0.0.0/0            /* default/my-udp-service: */ udp dpt:6000 redir ports 47497
    
    Chain KUBE-NODEPORT-HOST (1 references)
     pkts bytes target     prot opt in     out     source               destination
        0     0 DNAT       udp  --  *      *       0.0.0.0/0            0.0.0.0/0            /* default/opensips:sipu */ udp dpt:5060 to:10.0.1.215:40482
        0     0 DNAT       udp  --  *      *       0.0.0.0/0            0.0.0.0/0            /* default/my-udp-service: */ udp dpt:6000 to:10.0.1.215:47497

I can share more info if required.  I have tried to dig in my capacity, but now I don't know what to look therefore requesting some help here.
Thanks

**Comment by suryaveerc** [comment]
Just an update on this issue: I tested this with **TCP** and it works fine.

**Comment by thockin** [comment]
UDP through the proxy is pretty geared towards DNS - single-use connections
with very short timeouts.  You might need to crank the UDP timeout up
(available in kube v1.1) - see `--udp-timeout` on kube-proxy

or you can try the pure iptables proxy in kube v1.1

**Comment by suryaveerc** [comment]
Thanks for the suggestion.

As you suggested I switched to pure iptables proxy. Below is the kube-proxy.service, which I changed:
```
[Unit]
Description=Kubernetes Proxy
Documentation=https://github.com/GoogleCloudPlatform/kubernetes
Requires=setup-network-environment.service
After=setup-network-environment.service

[Service]
ExecStartPre=/usr/bin/curl -L -o /opt/bin/kube-proxy -z /opt/bin/kube-proxy https://storage.googleapis.com/kubernetes-release/release/v1.0.3/bin/linux/amd64/kube-proxy
ExecStartPre=/usr/bin/chmod +x /opt/bin/kube-proxy
# wait for kubernetes master to be up and ready
ExecStartPre=/opt/bin/wupiao 10.0.1.240 8080
ExecStart=/opt/bin/kube-proxy \
--master=10.0.1.240:8080 \
--logtostderr=true --proxy-mode="iptables"
Restart=always
RestartSec=10
```
 
Now, nothing is working. I mean I can't connect to service or nodeport. Pods are only accessible from the pod IPs.

I don't see change in iptables after creating services. sudo iptables -L -nvt nat.

```
core@minion02 ~ $ sudo iptables -L -vnt nat
Chain PREROUTING (policy ACCEPT 5 packets, 368 bytes)
 pkts bytes target     prot opt in     out     source               destination
    5   368 DOCKER     all  --  *      *       0.0.0.0/0            0.0.0.0/0            ADDRTYPE match dst-type LOCAL

Chain INPUT (policy ACCEPT 5 packets, 368 bytes)
 pkts bytes target     prot opt in     out     source               destination

Chain OUTPUT (policy ACCEPT 1248 packets, 78710 bytes)
 pkts bytes target     prot opt in     out     source               destination
    1    70 DOCKER     all  --  *      *       0.0.0.0/0           !127.0.0.0/8          ADDRTYPE match dst-type LOCAL

Chain POSTROUTING (policy ACCEPT 1246 packets, 78580 bytes)
 pkts bytes target     prot opt in     out     source               destination
    6   348 FLANNEL    all  --  *      *       10.244.0.0/16        0.0.0.0/0

Chain DOCKER (2 references)
 pkts bytes target     prot opt in     out     source               destination

Chain FLANNEL (1 references)
 pkts bytes target     prot opt in     out     source               destination
    3   200 ACCEPT     all  --  *      *       0.0.0.0/0            10.244.0.0/16
    0     0 MASQUERADE  all  --  *      *       0.0.0.0/0           !224.0.0.0/4
```

Thanks

**Comment by suryaveerc** [comment]
Hi,

I didn't noticed the second 200 OK message in the log posted above. But that tells the story.

The backgroud is, The SIP client and server both connects on port 5060. So, when server in the POD replies on port 5060, the request is intercepted by the kube-proxy and being redirect to the POD itself because of the iptable rules.

From above, log of POD :

**Actual request to POD**
***10.244.2.1.50867 > 10.244.2.4.5060***: [bad udp cksum 0x1ddc -> 0xe738!] SIP, length: 982
        PUBLISH sip:service-1@opensipstest.org SIP/2.0
**The sent response from POD**
 ***10.244.2.4.5060 > 10.244.2.1.5060***: [bad udp cksum 0x1b95 -> 0xeddc!] SIP, length: 399
        SIP/2.0 200 OK
**The redirected response to the POD**
***10.244.2.1.58836 > 10.244.2.4.5060***: [bad udp cksum 0x1b95 -> 0x1bcc!] SIP, length: 399
        SIP/2.0 200 OK

Is this the expected behaviour? Can we have some workaround for this?

Thanks

**Comment by thockin** [comment]
I am a little lost on this.  Can you restate the problem with a bit more
step-by-step detail?

Are the clients and servers all pods?  Why are you using a node port?

From your iptables output it doesn't look like kube-proxy is running - are
you sure it is?  check the logs.

**Comment by suryaveerc** [comment]
Hi,

I'll try to give as much detail as possible:

1. Installed K8s on Openstack following [K8s on CoreOS] (http://kubernetes.io/v1.1/docs/getting-started-guides/coreos/coreos_multinode_cluster.html ). 1 Master Node, 2 Minion Nodes.
2. Created few services (mysql, redis, tomcat, opensips).
3. Mysql, redis, tomcat are all accessed by Opensips.
4. Only Opensips is exposed to outside world.
5. Opensips listens on UDP port 5060 (SIP), that's why I created Nodeport service (I read that LoadBalancer does not support UDP).
6. From Opensips to other services everything is working as expected.
7. SIP clients also connects from port 5060, so they send request from port 5060.
8. Client is outside of Openstack but on same network.
9. Request reaches the Opensips pod and reply is sent from OpenSips to the IP from where it received the request and port 5060 **(I suspect the main issue is here, I think the port is wrong. A bit confused here, opened an issue with Opensips, no reply yet.)**
10. The response from pod is redirected to the pod.

Currently no access to the environment otherwise would have shared the logs of Kube-proxy.

Thanks,
Suryaveer

**Comment by thockin** [comment]
On Tue, Dec 29, 2015 at 5:09 PM, suryaveerc <notifications@github.com> wrote:

> Installed K8s on Openstack following K8s on CoreOS. 1 Master Node, 2 Minion Nodes.
> Created few services (mysql, redis, tomcat, opensips).
> Mysql, redis, tomcat are all accessed by Opensips.
> Only Opensips is exposed to outside world.
> Opensips listens on UDP port 5060 (SIP), that's why I created Nodeport service (I read that LoadBalancer does not support UDP).

Right, the UDP load-balancer PR is not merged yet.  Damn.  To do this
you must have changed the cluster's service node port range - it
defaults to 30000-32767, right?

> From Opensips to other services everything is working as expected.
> SIP clients also connects from port 5060, so they send request from port 5060.
> Client is outside of Openstack but on same network.

same IP space or is it crossing virtualization planes or rfc1918
edges?  I'm trying to avoid node port because I think I see what is
happening.  Let's see if we can piece it together.

Client C:5060 hits the node N:5060, so the packet is {s=C:5060,
d=N:5060}.  That gets masqueraded (all node ports do) as it is passed
to a pod P, so the packet should say {s=N:5060, d=P:5060}.  The
response will be {s=P:5060, d=N:5060}.  This seems like it could be a
problem if that gets trapped and looped back to the pod!

I tried to repro this but could not.  I spent 3 hours trying my best
to break it, but failed.  The kernel is smart enough to un-NAT before
applying rules, so the loop doesn't happen.

TL;DR: can't repro

> Request reaches the Opensips pod and reply is sent from OpenSips to the IP from where it received the request and port 5060 (I suspect the main issue is here, I think the port is wrong. A bit confused here, opened an issue with Opensips, no reply yet.)
> The response from pod is redirected to the pod.

Can you craft a simple repro that shows this - step-by-step exact
instructions with as little noise as possible (e.g. if mysql is not
required to repro, leave it out :)

**Comment by suryaveerc** [comment]
@thockin 

>Right, the UDP load-balancer PR is not merged yet.  Damn.  To do this
you must have changed the cluster's service node port range - it
defaults to 30000-32767, right?

Yes, I changed the defaults.
>same IP space or is it crossing virtualization planes or rfc1918
edges?  I'm trying to avoid node port because I think I see what is
happening.  Let's see if we can piece it together.

Clients are on same network as Openstack.

>I tried to repro this but could not.  I spent 3 hours trying my best
to break it, but failed.  The kernel is smart enough to un-NAT before
applying rules, so the loop doesn't happen.

Of course you couldn't repro it. As I said that I suspect the issue to be from Opensips and that was the case. K8s is working fine and Opensips too but there's a config in Opensips for such situation. Done that and now eerything is fine.

Actual request to POD
*10.244.2.1.50867 > 10.244.2.4.5060: [bad udp cksum 0x1ddc -> 0xe738!] SIP, length: 982*
PUBLISH sip:service-1@opensipstest.org SIP/2.0
The sent response from POD
**~~10.244.2.4.5060 > 10.244.2.1.5060: [bad udp cksum 0x1b95 -> 0xeddc!] SIP, length: 399~~**
SIP/2.0 200 OK
This has to be 
**10.244.2.4.5060 > 10.244.2.1.*50867*: [bad udp cksum 0x1b95 -> 0xeddc!] SIP, length: 399**
SIP/2.0 200 OK
The redirected response to the POD
**~~10.244.2.1.58836 > *10.244.2.4*.5060: [bad udp cksum 0x1b95 -> 0x1bcc!] SIP, length: 399~~**
and this
**10.244.2.1.58836 > *ClientIP*.5060: [bad udp cksum 0x1b95 -> 0x1bcc!] SIP, length: 399**
SIP/2.0 200 OK

In short, the response from the POD should have been sent to the IP:Port (kube proxy ip:port) the request is received from instead the response was sent to IP of Kubeproxy and some port(5060 in this case) that has rule set on Node. 

>Client C:5060 hits the node N:5060, so the packet is {s=C:5060,
d=N:5060}.  That gets masqueraded (all node ports do) as it is passed
to a pod P, so the packet should say **{s=N:5060, d=P:5060}** shouldn't it be **{s=*Kubeproxy port*, d=P:5060}**
.  The
response will be **{s=P:5060, d=N:5060}** and this **{s=P:5060, d=Kubeproxy port}** .  This seems like it could be a problem if that gets trapped and looped back to the pod!

New to K8s, tried to explain what I think must have happened and sorry for your 3 hrs. 

Issue can be closed now.

Thanks & Regards,
Suryaveer

**Comment by thockin** [comment]
I still don't understand what was happening - what you describe "the
response was sent to IP of Kubeproxy and some port(5060 in this case) that
has rule set on Node" should not trigger iptables until AFTER NAT is
resolved.

If you have a way to repro it , I'd like to try

**Comment by suryaveerc** [comment]
@thockin 
I can reproduce it anytime. Let me know what details you want to see.

Thanks.

**Comment by thockin** [comment]
I need the simplest step-by-step repro you can come up with, please.

Thanks!

Tim

**Comment by suryaveerc** [comment]
@thockin 

Wish you a very Happy New Year.

Sorry for the delay. Below is the simple scenario to reproduce it.

Application A listens on port 5060 running in a POD(P) exposed port 5060.
Nodeport service (N) created, exposed port 5060.
NAT rule to forward request to KubeProxy(KP) some random port (say 34567)
Client (C) connects from port 5060.
KP forwards to Pod from some another random port (say 45678)

**Request Flow:**
1. {C:5060, N:5060}  -- OK
2. {N:5060, KP:34567} -- OK
3. {KP:45678, P:5060} -- OK
**Response flow:**
1. {P:5060,KP:5060} --> should have been {P:5060, KP:45678} (Client port choosen instead of KP)
2. {KP:5060,P:5060} -- response sent to POD

So, basically what happens is that the application in POD sends the response to the IP of KP and port of actual client (*SIP message contains the IP:PORT information of the client.*)

Another scenario, client sends request from some random port say 6070. In this case the KP doesn't redirect the response back to POD instead it fails with ICMP port unreachable error. I guess its because no one is listening on the port 6070(*the client port*) on node.

**Request Flow:**
1. {C:6070, N:5060}  -- OK
2. {N:5060, KP:34567} -- OK
3. {KP:45678, P:5060} -- OK
**Response flow:**
1. {P:5060,KP:6070} --> should have been {P:5060, KP:45678}
2. ICMP error in Node.

From this what I understand that somehow iptables rules are getting triggered before KP can act. 

I hope, I made the steps to reproduce simple enough.

Thanks,
Suryaveer

**Comment by thockin** [comment]
On Fri, Jan 1, 2016 at 12:50 AM, suryaveerc <notifications@github.com> wrote:
>
> @thockin
>
> Wish you a very Happy New Year.
>
> Sorry for the delay. Below is the simple scenario to reproduce it.
>
> Application A listens on port 5060 running in a POD(P) exposed port 5060.
> Nodeport service (N) created, exposed port 5060.
> NAT rule to forward request to KubeProxy(KP) some random port (say 34567)
> Client (C) connects from port 5060.
> KP forwards to Pod from some another random port (say 45678)
>
> Request Flow:
> 1. {C:5060, N:5060} -- OK
> 2. {N:5060, KP:34567} -- OK
> 3. {KP:45678, P:5060} -- OK
> Response flow:
> 1. {P:5060,KP:5060} --> should have been {P:5060, KP:45678} (Client port choosen instead of KP)
> 2. {KP:5060,P:5060} -- response sent to POD
>
> So, basically what happens is that the application in POD sends the response to the IP of KP and port of actual client (SIP message contains the IP:PORT information of the client.)

Ahh, if the SIP message contains the ip:port, I can maybe see
problems.  The test I set up (which is pretty much what you described)
assumed simple UDP responses to the post-NAT IP, which gets un-NATed
upon receipt.

> Another scenario, client sends request from some random port say 6070. In this case the KP doesn't redirect the response back to POD instead it fails with ICMP port unreachable error. I guess its because no one is listening on the port 6070(the client port) on node.
>
> Request Flow:
> 1. {C:6070, N:5060} -- OK
> 2. {N:5060, KP:34567} -- OK
> 3. {KP:45678, P:5060} -- OK
> Response flow:
> 1. {P:5060,KP:6070} --> should have been {P:5060, KP:45678}
> 2. ICMP error in Node.
>
> From this what I understand that somehow iptables rules are getting triggered before KP can act.
>
> I hope, I made the steps to reproduce simple enough.

I was hoping for something like "run pod X; run pod Y; create a
service S; netcat X:5060 -p 5060; observe failure" :)

I'll try again to repro, but if my understanding is correct, and SIP
is taking port into the app layer, then yeah, it's trouble.

**Comment by suryaveerc** [comment]
@thockin 
I actually tried to do some simple C program to repro but stuck somewhere. May be I can do it sometime.

I have another issue, I can open a new one if you wish. 
The document says that service provide loadbalancing but my NODEPORT UDP service does not do this. If I send request at some intervals I see those going to different pods but if I send them continuously they go to same pod.

Thanks

**Comment by thockin** [comment]
UDP is well known to be hard to proxy because there isn't a "connection".
If we set timeouts too low then UDP "flows" get broken.  If we set timeouts
too long then you get this.  Once there is a NAT entry for a given
client:port to a particular backend:port, that will be reused.  UDP can not
tell the difference between another packet in a "flow" and a new connection.

If you use a different source port you should get a different backend,
though.  Is that not happening?

**Comment by suryaveerc** [comment]
@thockin 

The test client creates all request from single port. I'll try to find the option to send requests from different ports and update here.

Thanks

**Comment by thockin** [comment]
I don't think ANY sort of proxy could handle routing UDP from the same
client ip:port to different backends, short of timeouts.

**Comment by suryaveerc** [comment]
I used different ports for requests and now it's working as expected.

Thanks
